--- linux-2.6.18.orig/security/apparmor/lsm.c
+++ linux-2.6.18/security/apparmor/lsm.c
@@ -199,9 +199,9 @@
 	return cap_netlink_send(sk, skb);
 }
 
-static int subdomain_netlink_recv(struct sk_buff *skb, int cap)
+static int subdomain_netlink_recv(struct sk_buff *skb)
 {
-	return cap_netlink_recv(skb, cap);
+	return cap_netlink_recv(skb);
 }
 
 static void subdomain_bprm_apply_creds(struct linux_binprm *bprm, int unsafe)
