Index: b/security/apparmor/apparmorfs.c
===================================================================
--- a/security/apparmor/apparmorfs.c
+++ b/security/apparmor/apparmorfs.c
@@ -86,14 +86,14 @@ static struct root_entry {
 	int parent_index;
 } root_entries[] = {
 	/* our root, normally /sys/kernel/security/apparmor */
-	{SECFS_AA, 	S_IFDIR, 0550},	/* DO NOT EDIT/MOVE */
+	{SECFS_AA, 	S_IFDIR, 0555},	/* DO NOT EDIT/MOVE */
 
 	/* interface for obtaining list of profiles currently loaded */
 	{"profiles", 	S_IFREG, 0440, &apparmorfs_profiles_fops,
 				       NULL},
 
 	/* interface for obtaining matching features supported */
-	{"matching",  	S_IFREG, 0440, &apparmorfs_matching_fops,
+	{"matching",  	S_IFREG, 0444, &apparmorfs_matching_fops,
 				       NULL},
 
 	/* interface for loading/removing/replacing profiles */
