Index: linux-2.6/security/Kconfig
===================================================================
--- linux-2.6.orig/security/Kconfig
+++ linux-2.6/security/Kconfig
@@ -94,6 +94,7 @@ config SECURITY_ROOTPLUG
 	  If you are unsure how to answer this question, answer N.
 
 source security/selinux/Kconfig
+source security/apparmor/Kconfig
 
 endmenu
 
Index: linux-2.6/security/Makefile
===================================================================
--- linux-2.6.orig/security/Makefile
+++ linux-2.6/security/Makefile
@@ -4,6 +4,7 @@
 
 obj-$(CONFIG_KEYS)			+= keys/
 subdir-$(CONFIG_SECURITY_SELINUX)	+= selinux
+subdir-$(CONFIG_SECURITY_APPARMOR)	+= apparmor
 
 # if we don't select a security model, use the default capabilities
 ifneq ($(CONFIG_SECURITY),y)
@@ -14,5 +15,8 @@ endif
 obj-$(CONFIG_SECURITY)			+= security.o dummy.o inode.o
 # Must precede capability.o in order to stack properly.
 obj-$(CONFIG_SECURITY_SELINUX)		+= selinux/built-in.o
+ifeq ($(CONFIG_SECURITY_APPARMOR),y)
+obj-$(CONFIG_SECURITY_APPARMOR)		+= apparmor/built-in.o
+endif
 obj-$(CONFIG_SECURITY_CAPABILITIES)	+= commoncap.o capability.o
 obj-$(CONFIG_SECURITY_ROOTPLUG)		+= commoncap.o root_plug.o
