---
 security/apparmor/lsm.c |   23 ++---------------------
 1 file changed, 2 insertions(+), 21 deletions(-)

--- a/security/apparmor/lsm.c
+++ b/security/apparmor/lsm.c
@@ -616,18 +616,8 @@ static int apparmor_setprocattr(struct t
 		return -EINVAL;
 
 	if (strcmp(command, "changehat") == 0) {
-		if (current != task) {
-			AA_WARN(GFP_KERNEL,
-				"Attempt by foreign task %s(%d) "
-				"[user %d] to changehat of task %s(%d)",
-				current->comm,
-				current->pid,
-				current->uid,
-				task->comm,
-				task->pid);
-
+		if (current != task)
 			return -EACCES;
-		}
 		error = aa_setprocattr_changehat(args);
 	} else if (strcmp(command, "setprofile")) {
 		struct aa_profile *profile;
@@ -636,17 +626,8 @@ static int apparmor_setprocattr(struct t
 		 * may change the profile of another task.
 		 */
 
-		if (!capable(CAP_SYS_ADMIN)) {
-			AA_WARN(GFP_KERNEL,
-				"Unprivileged attempt by task %s(%d) "
-				"[user %d] to assign profile to task %s(%d)",
-				current->comm,
-				current->pid,
-				current->uid,
-				task->comm,
-				task->pid);
+		if (!capable(CAP_SYS_ADMIN))
 			return -EACCES;
-		}
 
 		profile = aa_get_profile(current);
 		if (profile) {
