---
 security/apparmor/lsm.c |   73 +++++-------------------------------------------
 1 file changed, 9 insertions(+), 64 deletions(-)

--- a/security/apparmor/lsm.c
+++ b/security/apparmor/lsm.c
@@ -133,30 +133,6 @@ static int apparmor_ptrace(struct task_s
 	return error;
 }
 
-static int apparmor_capget(struct task_struct *task,
-			    kernel_cap_t *effective,
-			    kernel_cap_t *inheritable,
-			    kernel_cap_t *permitted)
-{
-	return cap_capget(task, effective, inheritable, permitted);
-}
-
-static int apparmor_capset_check(struct task_struct *task,
-				  kernel_cap_t *effective,
-				  kernel_cap_t *inheritable,
-				  kernel_cap_t *permitted)
-{
-	return cap_capset_check(task, effective, inheritable, permitted);
-}
-
-static void apparmor_capset_set(struct task_struct *task,
-				 kernel_cap_t *effective,
-				 kernel_cap_t *inheritable,
-				 kernel_cap_t *permitted)
-{
-	cap_capset_set(task, effective, inheritable, permitted);
-}
-
 static int apparmor_capable(struct task_struct *task, int cap)
 {
 	int error;
@@ -188,26 +164,6 @@ static int apparmor_sysctl(struct ctl_ta
 	return error;
 }
 
-static int apparmor_syslog(int type)
-{
-	return cap_syslog(type);
-}
-
-static int apparmor_netlink_send(struct sock *sk, struct sk_buff *skb)
-{
-	return cap_netlink_send(sk, skb);
-}
-
-static int apparmor_netlink_recv(struct sk_buff *skb, int cap)
-{
-	return cap_netlink_recv(skb, cap);
-}
-
-static void apparmor_bprm_apply_creds(struct linux_binprm *bprm, int unsafe)
-{
-	cap_bprm_apply_creds(bprm, unsafe);
-}
-
 static int apparmor_bprm_set_security(struct linux_binprm *bprm)
 {
 	/* handle capability bits with setuid, etc */
@@ -594,17 +550,6 @@ static void apparmor_task_free_security(
 	aa_release(task);
 }
 
-static int apparmor_task_post_setuid(uid_t id0, uid_t id1, uid_t id2,
-				     int flags)
-{
-	return cap_task_post_setuid(id0, id1, id2, flags);
-}
-
-static void apparmor_task_reparent_to_init(struct task_struct *task)
-{
-	cap_task_reparent_to_init(task);
-}
-
 static int apparmor_getprocattr(struct task_struct *task, char *name,
 				char **value)
 {
@@ -696,17 +641,17 @@ static int apparmor_setprocattr(struct t
 
 struct security_operations apparmor_ops = {
 	.ptrace =			apparmor_ptrace,
-	.capget =			apparmor_capget,
-	.capset_check =			apparmor_capset_check,
-	.capset_set =			apparmor_capset_set,
+	.capget =			cap_capget,
+	.capset_check =			cap_capset_check,
+	.capset_set =			cap_capset_set,
 	.sysctl =			apparmor_sysctl,
 	.capable =			apparmor_capable,
-	.syslog =			apparmor_syslog,
+	.syslog =			cap_syslog,
 
-	.netlink_send =			apparmor_netlink_send,
-	.netlink_recv =			apparmor_netlink_recv,
+	.netlink_send =			cap_netlink_send,
+	.netlink_recv =			cap_netlink_recv,
 
-	.bprm_apply_creds =		apparmor_bprm_apply_creds,
+	.bprm_apply_creds =		cap_bprm_apply_creds,
 	.bprm_set_security =		apparmor_bprm_set_security,
 	.bprm_secureexec =		apparmor_bprm_secureexec,
 
@@ -736,8 +681,8 @@ struct security_operations apparmor_ops 
 	.task_create =			apparmor_task_create,
 	.task_alloc_security =		apparmor_task_alloc_security,
 	.task_free_security =		apparmor_task_free_security,
-	.task_post_setuid =		apparmor_task_post_setuid,
-	.task_reparent_to_init =	apparmor_task_reparent_to_init,
+	.task_post_setuid =		cap_task_post_setuid,
+	.task_reparent_to_init =	cap_task_reparent_to_init,
 
 	.getprocattr =			apparmor_getprocattr,
 	.setprocattr =			apparmor_setprocattr,
