Index: b/security/apparmor/lsm.c
===================================================================
--- a/security/apparmor/lsm.c
+++ b/security/apparmor/lsm.c
@@ -207,7 +207,7 @@ static int apparmor_sb_mount(char *dev_n
 
 static int apparmor_umount(struct vfsmount *mnt, int flags)
 {
-	return aa_reject_syscall(current, GFP_ATOMIC, "umount");
+	return aa_reject_syscall(current, GFP_KERNEL, "umount");
 }
 
 static int apparmor_inode_mkdir(struct inode *dir, struct dentry *dentry,
