---
 security/apparmor/lsm.c |    3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

--- a/security/apparmor/lsm.c
+++ b/security/apparmor/lsm.c
@@ -366,8 +366,7 @@ static inline int aa_mask_permissions(in
 static int apparmor_inode_create(struct inode *dir, struct dentry *dentry,
 				 struct vfsmount *mnt, int mask)
 {
-	/* FIXME: may move to MAY_APPEND later */
-	return aa_permission("inode_create", dir, dentry, mnt, MAY_WRITE, 0);
+	return aa_permission("inode_create", dir, dentry, mnt, MAY_APPEND, 0);
 }
 
 static int apparmor_inode_link(struct dentry *old_dentry,
