---
 security/apparmor/lsm.c |   19 ++++++++++++++-----
 1 file changed, 14 insertions(+), 5 deletions(-)

--- a/security/apparmor/lsm.c
+++ b/security/apparmor/lsm.c
@@ -114,20 +114,29 @@ static int apparmor_ptrace(struct task_s
 	/**
 	 * parent can ptrace child when
 	 * - parent is unconfined
-	 * - parent is in complain mode
-	 * - parent and child are confined by the same profile
+	 * - parent & child are in the same namespace &&
+	 *   - parent is in complain mode
+	 *   - parent and child are confined by the same profile
+	 *   - parent profile has CAP_SYS_PTRACE
 	 */
 
 	rcu_read_lock();
 	cxt = aa_task_context(parent);
 	child_cxt = aa_task_context(child);
 	child_profile = child_cxt ? child_cxt->profile : NULL;
-	error = aa_may_ptrace(cxt, child_profile);
-	if (cxt && PROFILE_COMPLAIN(cxt->profile)) {
-		LOG_HINT(cxt->profile, GFP_ATOMIC, HINT_PTRACE,
+	if (cxt && (parent->nsproxy != child->nsproxy)) {
+		AA_WARN(GFP_ATOMIC,
+			"REJECTING ptrace across namespace of %d by %d",
+			parent->pid, child->pid);
+		error = -EPERM;
+	} else {
+		error = aa_may_ptrace(cxt, child_profile);
+		if (cxt && PROFILE_COMPLAIN(cxt->profile)) {
+			LOG_HINT(cxt->profile, GFP_ATOMIC, HINT_PTRACE,
 				 "pid=%d child=%d\n",
 				 current->pid, child->pid);
 		}
+	}
 	rcu_read_unlock();
 
 	return error;
