---
 security/apparmor/main.c |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

--- a/security/apparmor/main.c
+++ b/security/apparmor/main.c
@@ -176,8 +176,10 @@ static int aa_audit_base(struct aa_profi
 		return type == AUDIT_APPARMOR_ALLOWED ? 0 : -ENOMEM;
 	}
 
+	audit_log_format(ab, " type=%d", type);
+
 	if (sa->operation)
-		audit_log_format(ab, "operation=\"%s\"", sa->operation);
+		audit_log_format(ab, " operation=\"%s\"", sa->operation);
 
 	if (sa->info) {
 		audit_log_format(ab, " info=\"%s\"", sa->info);
