---
 security/apparmor/main.c |    3 +++
 1 file changed, 3 insertions(+)

--- a/security/apparmor/main.c
+++ b/security/apparmor/main.c
@@ -405,6 +405,9 @@ static int aa_audit_base(struct aa_profi
 	if (sa->denied_mask)
 		aa_audit_file_mask(ab, "denied_mask", sa->denied_mask);
 
+	if (sa->request_mask)
+		audit_log_format(ab, " fsuid=%d", current->fsuid);
+
 	if (sa->iattr) {
 		struct iattr *iattr = sa->iattr;
 
