---
 security/apparmor/main.c |   10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

--- a/security/apparmor/main.c
+++ b/security/apparmor/main.c
@@ -331,10 +331,10 @@ static int aa_audit_base(struct aa_profi
 	}
 
 	if (sa->task)
-		audit_log_format(ab, " task=\"%d\"", sa->task);
+		audit_log_format(ab, " task=%d", sa->task);
 
 	if (sa->parent)
-		audit_log_format(ab, "parent=\"%d\"", sa->parent);
+		audit_log_format(ab, " parent=%d", sa->parent);
 
 	if (sa->name) {
 		audit_log_format(ab, " name=");
@@ -347,9 +347,9 @@ static int aa_audit_base(struct aa_profi
 	}
 
 	if (sa->magic_token)
-		audit_log_format(ab, " magic_token=\"%llu\"", sa->magic_token);
+		audit_log_format(ab, " magic_token=%llu", sa->magic_token);
 
-	audit_log_format(ab, "pid=\"%d\"", current->pid);
+	audit_log_format(ab, " pid=%d", current->pid);
 
 	if (profile) {
 		audit_log_format(ab, " profile=");
@@ -854,7 +854,7 @@ repeat:
 			goto repeat;
 		if (PTR_ERR(old_profile) == -EPERM) {
 			sa.denied_mask = MAY_EXEC;
-			sa.info = "Unable to set profile due to ptrace";
+			sa.info = "unable to set profile due to ptrace";
 			sa.task = current->parent->pid;
 			aa_audit_reject(profile, &sa);
 		}
