---
 security/apparmor/lsm.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/security/apparmor/lsm.c
+++ b/security/apparmor/lsm.c
@@ -633,7 +633,7 @@ static int apparmor_setprocattr(struct t
 		if (current != task)
 			return -EACCES;
 		error = aa_setprocattr_changeprofile(args);
-	} else if (strcmp(command, "setprofile")) {
+	} else if (strcmp(command, "setprofile") == 0) {
 		struct aa_profile *profile;
 
 		/* Only an unconfined process with admin capabilities
