---
 security/apparmor/main.c |   25 ++++++++++++-------------
 1 file changed, 12 insertions(+), 13 deletions(-)

--- a/security/apparmor/main.c
+++ b/security/apparmor/main.c
@@ -937,7 +937,7 @@ repeat:
 static int do_change_profile(struct aa_profile *expected, const char *name,
 			     u64 cookie, int restore, struct aa_audit *sa)
 {
-	struct aa_profile *profile = NULL, *old_profile = NULL,
+	struct aa_profile *new_profile = NULL, *old_profile = NULL,
 		*previous_profile = NULL;
 	struct aa_task_context *new_cxt, *cxt;
 	int error = 0;
@@ -948,23 +948,23 @@ static int do_change_profile(struct aa_p
 	if (!new_cxt)
 		return -ENOMEM;
 
-	profile = aa_find_profile(name);
-	if (!profile && !restore) {
+	new_profile = aa_find_profile(name);
+	if (!new_profile && !restore) {
 		if (!PROFILE_COMPLAIN(expected)) {
 			error = -ENOENT;
 			goto out;
 		}
-		profile = aa_dup_profile(null_complain_profile);
+		new_profile = aa_dup_profile(null_complain_profile);
 	}
 
-	cxt = lock_task_and_profiles(current, profile);
+	cxt = lock_task_and_profiles(current, new_profile);
 	if (!cxt) {
 		error = -EPERM;
 		goto out;
 	}
 	old_profile = cxt->profile;
 
-	if (cxt->profile != expected || (profile && profile->isstale)) {
+	if (cxt->profile != expected || (new_profile && new_profile->isstale)) {
 		error = -ESTALE;
 		goto out;
 	}
@@ -984,34 +984,33 @@ static int do_change_profile(struct aa_p
 	} else
 		previous_profile = cxt->profile;
 
-	if ((current->ptrace & PT_PTRACED) && aa_may_ptrace(cxt, profile)) {
+	if ((current->ptrace & PT_PTRACED) && aa_may_ptrace(cxt, new_profile)) {
 		error = -EACCES;
 		goto out;
 	}
 
-	if (profile == null_complain_profile)
+	if (new_profile == null_complain_profile)
 		aa_audit_hint(cxt->profile, sa);
 
 	if (APPARMOR_AUDIT(cxt))
 		aa_audit_message(cxt->profile, sa, AUDIT_APPARMOR_AUDIT);
 
 	if (!restore && cookie)
-		aa_change_task_context(current, new_cxt, profile, cookie,
+		aa_change_task_context(current, new_cxt, new_profile, cookie,
 				       previous_profile);
 	else
 		/* either return to previous_profile, or a permanent change */
-		aa_change_task_context(current, new_cxt, profile, 0, NULL);
+		aa_change_task_context(current, new_cxt, new_profile, 0, NULL);
 
 out:
 	if (aa_task_context(current) != new_cxt)
 		aa_free_task_context(new_cxt);
 	task_unlock(current);
-	unlock_both_profiles(old_profile, profile);
-	aa_put_profile(profile);
+	unlock_both_profiles(old_profile, new_profile);
+	aa_put_profile(new_profile);
 	return error;
 }
 
-
 /**
  * aa_change_profile - change profile to/from previous stored profile
  * @name: name of profile to change to
