Index: b/security/apparmor/main.c
===================================================================
--- a/security/apparmor/main.c
+++ b/security/apparmor/main.c
@@ -649,7 +649,8 @@ int aa_perm_dir(struct aa_profile *profi
 	sa.flags = 0;
 	sa.gfp_mask = GFP_KERNEL;
 
-	return aa_perm_dentry(profile, dentry, mnt, &sa, mask, AA_CHECK_DIR);
+	return aa_perm_dentry(profile, dentry, mnt, &sa, mask,
+			      AA_CHECK_DIR | AA_CHECK_LEAF);
 }
 
 /**
