---
 security/apparmor/apparmor.h |   43 ++++++++++++++++++++++++++++---------------
 1 file changed, 28 insertions(+), 15 deletions(-)

--- a/security/apparmor/apparmor.h
+++ b/security/apparmor/apparmor.h
@@ -26,25 +26,33 @@
 #define AA_MAY_LINK			0x0010
 #define AA_MAY_LOCK			0x0020
 #define AA_EXEC_MMAP			0x0040
-#define AA_EXEC_UNSAFE			0x0080
-#define AA_EXEC_MOD_0			0x0100
-#define AA_EXEC_MOD_1			0x0200
+#define AA_MAY_MOUNT			0x0080	/* no direct audit mapping */
+#define AA_EXEC_UNSAFE			0x0100
+#define AA_EXEC_MOD_0			0x0200
+#define AA_EXEC_MOD_1			0x0400
+#define AA_EXEC_MOD_2			0x0800
+#define AA_EXEC_MOD_3			0x1000
+#define AA_EXEC_MOD_4			0x2000
+
 #define AA_BASE_PERMS			(MAY_READ | MAY_WRITE | MAY_EXEC | \
 					 MAY_APPEND | AA_MAY_LINK | \
 					 AA_MAY_LOCK | AA_EXEC_MMAP | \
-					 AA_EXEC_UNSAFE | AA_EXEC_MOD_0 | \
-					 AA_EXEC_MOD_1)
-#define AA_LINK_SUBSET_TEST		0x0020
-
-#define AA_EXEC_UNCONFINED		0
-#define AA_EXEC_INHERIT			AA_EXEC_MOD_0
-#define AA_EXEC_PROFILE			AA_EXEC_MOD_1
-#define AA_EXEC_PIX			(AA_EXEC_MOD_0 | AA_EXEC_MOD_1)
-
-#define AA_EXEC_MODIFIERS		(AA_EXEC_MOD_0 | AA_EXEC_MOD_1)
+					 AA_MAY_MOUNT | AA_EXEC_UNSAFE | \
+					 AA_EXEC_MOD_0 | AA_EXEC_MOD_1 | \
+					 AA_EXEC_MOD_2 | AA_EXEC_MOD_3 | \
+					 AA_EXEC_MOD_4)
+
+#define AA_EXEC_UNCONFINED		AA_EXEC_MOD_0
+#define AA_EXEC_INHERIT			AA_EXEC_MOD_1
+#define AA_EXEC_PROFILE			(AA_EXEC_MOD_0 | AA_EXEC_MOD_1)
+#define AA_EXEC_PIX			 AA_EXEC_MOD_2
+
+#define AA_EXEC_MODIFIERS		(AA_EXEC_MOD_0 | AA_EXEC_MOD_1 | \
+					 AA_EXEC_MOD_2 | AA_EXEC_MOD_3 | \
+					 AA_EXEC_MOD_4)
 
 #define AA_USER_SHIFT			0
-#define AA_OTHER_SHIFT			10
+#define AA_OTHER_SHIFT			14
 
 #define AA_USER_PERMS			(AA_BASE_PERMS << AA_USER_SHIFT)
 #define AA_OTHER_PERMS			(AA_BASE_PERMS << AA_OTHER_SHIFT)
@@ -68,11 +76,16 @@
 #define AA_ALL_EXEC_MODS		(AA_USER_EXEC_MODS | \
 					 AA_OTHER_EXEC_MODS)
 
+/* overloaded permissions for link pairs */
+#define AA_LINK_SUBSET_TEST		0x0020
+
 /* shared permissions that are not duplicated in user::other */
+#define AA_AUDIT_FIELD			0x10000000
 #define AA_CHANGE_HAT			0x20000000
 #define AA_CHANGE_PROFILE		0x40000000
 
-#define AA_SHARED_PERMS			(AA_CHANGE_HAT | AA_CHANGE_PROFILE)
+#define AA_SHARED_PERMS			(AA_CHANGE_HAT | AA_CHANGE_PROFILE | \
+					 AA_AUDIT_FIELD)
 
 #define AA_VALID_PERM_MASK		(AA_FILE_PERMS | AA_SHARED_PERMS)
 
