---
 security/apparmor/main.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/security/apparmor/main.c
+++ b/security/apparmor/main.c
@@ -338,7 +338,7 @@ static void aa_audit_file_mask(struct au
 	aa_audit_file_sub_mask(ab, other,
 			       (mask & AA_OTHER_PERMS) >> AA_OTHER_SHIFT);
 
-	audit_log_format(ab, " %s=\"%s::%s\"", name, user, other);
+	audit_log_format(ab, " %s=\"%s:%s\"", name, user, other);
 }
 
 static const char *address_families[] = {
