---
 security/apparmor/main.c |    4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

--- a/security/apparmor/main.c
+++ b/security/apparmor/main.c
@@ -1239,9 +1239,7 @@ repeat:
 		return -ENOENT;
 	}
 
-	if (PROFILE_COMPLAIN(profile) ||
-	    (ns == profile->ns &&
-	     (aa_match(profile->file_rules, name) & AA_CHANGE_PROFILE)))
+	if (PROFILE_COMPLAIN(profile))
 		error = do_change_profile(profile, ns, name, 0, 0, &sa);
 	else {
 		/* check for a rule with a namespace prepended */
