Index: b/security/apparmor/main.c
===================================================================
--- a/security/apparmor/main.c
+++ b/security/apparmor/main.c
@@ -198,7 +198,11 @@ static int aa_perm_dentry(struct aa_prof
 	sa->name = aa_get_name(dentry, mnt, &buffer, is_dir);
 
 	if (IS_ERR(sa->name)) {
-		denied_mask = PTR_ERR(sa->name);
+		/* deleted files are given a pass on permission checks */
+		if (PTR_ERR(sa->name) == -ENOENT)
+			denied_mask = 0;
+		else
+			denied_mask = PTR_ERR(sa->name);
 		sa->name = NULL;
 	} else {
 		denied_mask = aa_file_denied(profile, sa->name, mask);
