---
 security/apparmor/lsm.c |   18 ------------------
 1 file changed, 18 deletions(-)

--- a/security/apparmor/lsm.c
+++ b/security/apparmor/lsm.c
@@ -717,22 +717,6 @@ static int apparmor_socket_shutdown(stru
 	return aa_revalidate_sk(sk, "socket_shutdown");
 }
 
-static int apparmor_socket_getpeersec_stream(struct socket *sock,
-			char __user *optval, int __user *optlen, unsigned len)
-{
-	struct sock *sk = sock->sk;
-
-	return aa_revalidate_sk(sk, "socket_getpeersec_stream");
-}
-
-static int apparmor_socket_getpeersec_dgram(struct socket *sock,
-					struct sk_buff *skb, u32 *secid)
-{
-	struct sock *sk = sock->sk;
-
-	return aa_revalidate_sk(sk, "socket_getpeersec_dgram");
-}
-
 static int apparmor_getprocattr(struct task_struct *task, char *name,
 				char **value)
 {
@@ -882,8 +866,6 @@ struct security_operations apparmor_ops 
 	.socket_getsockopt =		apparmor_socket_getsockopt,
 	.socket_setsockopt =		apparmor_socket_setsockopt,
 	.socket_shutdown =		apparmor_socket_shutdown,
-	.socket_getpeersec_stream =	apparmor_socket_getpeersec_stream,
-	.socket_getpeersec_dgram =	apparmor_socket_getpeersec_dgram,
 };
 
 static void info_message(const char *str)
