Index: b/security/apparmor/main.c
===================================================================
--- a/security/apparmor/main.c
+++ b/security/apparmor/main.c
@@ -844,16 +844,16 @@ repeat:
 		 */
 		exec_mode = aa_match(profile->file_rules, filename);
 
-		if (exec_mode & AA_EXEC_MODIFIERS) {
-			switch (exec_mode & AA_EXEC_MODIFIERS) {
-			case AA_EXEC_INHERIT:
+		if (exec_mode & (MAY_EXEC | AA_EXEC_MODIFIERS)) {
+			switch (exec_mode & (MAY_EXEC | AA_EXEC_MODIFIERS)) {
+			case MAY_EXEC | AA_EXEC_INHERIT:
 				AA_DEBUG("%s: INHERIT %s\n",
 					 __FUNCTION__,
 					 filename);
 				/* nothing to be done here */
 				goto cleanup;
 
-			case AA_EXEC_UNCONSTRAINED:
+			case MAY_EXEC | AA_EXEC_UNCONSTRAINED:
 				AA_DEBUG("%s: UNCONSTRAINED %s\n",
 					 __FUNCTION__,
 					 filename);
@@ -862,7 +862,7 @@ repeat:
 				newprofile = NULL;
 				break;
 
-			case AA_EXEC_PROFILE:
+			case MAY_EXEC | AA_EXEC_PROFILE:
 				AA_DEBUG("%s: PROFILE %s\n",
 					 __FUNCTION__,
 					 filename);
