Index: b/security/apparmor/procattr.c
===================================================================
--- a/security/apparmor/procattr.c
+++ b/security/apparmor/procattr.c
@@ -40,6 +40,7 @@ int aa_getprocattr(struct aa_profile *pr
 		memcpy(str, mode_str, strlen(mode_str));
 		str += strlen(mode_str);
 		*str++ = '\n';
+		str -= *len;
 	} else {
 		const char *unconfined_str = "unconfined\n";
 
