---
 security/apparmor/lsm.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/security/apparmor/lsm.c
+++ b/security/apparmor/lsm.c
@@ -112,7 +112,7 @@ static int apparmor_ptrace(struct task_s
 	struct aa_profile *child_profile;
 	int error = 0;
 
-	/**
+	/*
 	 * parent can ptrace child when
 	 * - parent is unconfined
 	 * - parent & child are in the same namespace &&
@@ -766,7 +766,7 @@ static void __exit apparmor_exit(void)
 
 	free_null_complain_profile();
 
-	/**
+	/*
 	 * Delay for an rcu cycle to make sure that all active task
 	 * context readers have finished, and all profiles have been
 	 * freed by their rcu callbacks.
