Index: b/security/apparmor/main.c
===================================================================
--- a/security/apparmor/main.c
+++ b/security/apparmor/main.c
@@ -1281,6 +1281,8 @@ void aa_change_task_context(struct task_
 		call_rcu(&old_cxt->rcu, free_aa_task_context_rcu_callback);
 	}
 	if (new_cxt) {
+		/* clear the caps_logged cache, so that new profile/hat has
+		 * chance to emit its own set of cap messages */
 		new_cxt->caps_logged = CAP_EMPTY_SET;
 		new_cxt->hat_magic = hat_magic;
 		new_cxt->task = task;
