Index: b/security/apparmor/module_interface.c
===================================================================
--- a/security/apparmor/module_interface.c
+++ b/security/apparmor/module_interface.c
@@ -381,9 +381,9 @@ static struct aa_profile *aa_activate_pr
 	AA_READ_X(e, AA_DYN_STRING, &profile->name, NULL);
 
 	error_string = "Invalid flags";
-	/* per profile debug flags (debug, complain, audit) */
+	/* per profile debug flags (complain, audit) */
 	AA_READ_X(e, AA_STRUCT, NULL, "flags");
-	AA_READ_X(e, AA_U32, &(profile->flags.debug), NULL);
+	AA_READ_X(e, AA_U32, NULL, NULL);  /* obsolete flags.debug */
 	AA_READ_X(e, AA_U32, &(profile->flags.complain), NULL);
 	AA_READ_X(e, AA_U32, &(profile->flags.audit), NULL);
 	AA_READ_X(e, AA_STRUCTEND, NULL, NULL);
Index: b/security/apparmor/apparmor.h
===================================================================
--- a/security/apparmor/apparmor.h
+++ b/security/apparmor/apparmor.h
@@ -109,7 +109,6 @@ struct aa_profile {
 	struct list_head list;
 	struct list_head sub;
 	struct {
-		int debug;
 		int complain;
 		int audit;
 	} flags;
