---
 security/apparmor/apparmor.h |   10 ----------
 1 file changed, 10 deletions(-)

--- a/security/apparmor/apparmor.h
+++ b/security/apparmor/apparmor.h
@@ -159,16 +159,6 @@ struct aa_audit {
 	int error_code;
 };
 
-/* audit types */
-#define AA_AUDITTYPE_FILE	1
-#define AA_AUDITTYPE_DIR	2
-#define AA_AUDITTYPE_ATTR	3
-#define AA_AUDITTYPE_XATTR	4
-#define AA_AUDITTYPE_LINK	5
-#define AA_AUDITTYPE_CAP	6
-#define AA_AUDITTYPE_MSG	7
-#define AA_AUDITTYPE_SYSCALL	8
-
 /* Flags for the permission check functions */
 #define AA_CHECK_FD	1  /* coming from a file descriptor */
 #define AA_CHECK_DIR	2  /* file type is directory */
