Index: linux-2.6/security/apparmor/lsm.c
===================================================================
--- linux-2.6.orig/security/apparmor/lsm.c
+++ linux-2.6/security/apparmor/lsm.c
@@ -256,10 +256,14 @@ static int aa_permission(struct inode *i
 	int error = 0;
 
 	if (mnt && mediated_filesystem(inode)) {
-		struct aa_profile *profile = aa_get_profile(current);
+		struct aa_profile *profile;
 
-		if (profile)
+		profile = aa_get_profile(current);
+		if (profile) {
+			if (inode && S_ISDIR(inode->i_mode))
+				check |= AA_CHECK_DIR;
 			error = aa_perm(profile, dentry, mnt, mask, check);
+		}
 		aa_put_profile(profile);
 	}
 	return error;
@@ -329,13 +333,19 @@ static int apparmor_inode_rename(struct 
 	profile = aa_get_profile(current);
 
 	if (profile) {
+		struct inode *inode = old_dentry->d_inode;
+		int check = AA_CHECK_LEAF;
+
+		if (inode && S_ISDIR(inode->i_mode))
+			check |= AA_CHECK_DIR;
 		if (old_mnt)
 			error = aa_perm(profile, old_dentry, old_mnt,
-					MAY_READ | MAY_WRITE, AA_CHECK_LEAF);
+					MAY_READ | MAY_WRITE, check);
 
-		if (!error && new_mnt)
+		if (!error && new_mnt) {
 			error = aa_perm(profile, new_dentry, new_mnt,
-					MAY_WRITE, AA_CHECK_LEAF);
+					MAY_WRITE, check);
+		}
 	}
 
 	aa_put_profile(profile);
Index: linux-2.6/security/apparmor/main.c
===================================================================
--- linux-2.6.orig/security/apparmor/main.c
+++ linux-2.6/security/apparmor/main.c
@@ -596,13 +596,9 @@ int aa_perm_xattr(struct aa_profile *pro
 int aa_perm(struct aa_profile *profile, struct dentry *dentry,
 	    struct vfsmount *mnt, int mask, int check)
 {
-	struct inode *inode = dentry->d_inode;
 	struct aa_audit sa;
 	int error = 0;
 
-	if (inode && S_ISDIR(inode->i_mode))
-		check |= AA_CHECK_DIR;
-
 	if ((check & (AA_CHECK_DIR | AA_CHECK_LEAF)) == AA_CHECK_DIR) {
 		/*
 		 * If checking a non-leaf directory, allow traverse and
