Index: linux-2.6.19.1/fs/open.c
===================================================================
--- linux-2.6.19.1.orig/fs/open.c
+++ linux-2.6.19.1/fs/open.c
@@ -483,6 +483,10 @@ asmlinkage long sys_chroot(const char __
 	if (!capable(CAP_SYS_CHROOT))
 		goto dput_and_out;
 
+	error = security_chroot(&nd);
+	if (error)
+		goto dput_and_out;
+
 	set_fs_root(current->fs, nd.mnt, nd.dentry);
 	set_fs_altroot();
 	error = 0;
Index: linux-2.6.19.1/include/linux/security.h
===================================================================
--- linux-2.6.19.1.orig/include/linux/security.h
+++ linux-2.6.19.1/include/linux/security.h
@@ -247,6 +247,9 @@ struct request_sock;
  *	Update module state after a successful pivot.
  *	@old_nd contains the nameidata structure for the old root.
  *      @new_nd contains the nameidata structure for the new root.
+ * @sb_chroot:
+ *	Check permission before chroot to chroot to point named by @nd
+ *	@nd contains the nameidata object for the new root
  *
  * Security hooks for inode operations.
  *
@@ -1219,6 +1222,7 @@ struct security_operations {
 			     struct nameidata * new_nd);
 	void (*sb_post_pivotroot) (struct nameidata * old_nd,
 				   struct nameidata * new_nd);
+	int (*sb_chroot) (struct nameidata * nd);
 
 	int (*inode_alloc_security) (struct inode *inode);	
 	void (*inode_free_security) (struct inode *inode);
@@ -1617,6 +1621,11 @@ static inline void security_sb_post_pivo
 	security_ops->sb_post_pivotroot (old_nd, new_nd);
 }
 
+static inline int security_chroot (struct nameidata *nd)
+{
+	return security_ops->sb_chroot (nd);
+}
+
 static inline int security_inode_alloc (struct inode *inode)
 {
 	inode->i_security = NULL;
@@ -2367,6 +2376,11 @@ static inline void security_sb_post_pivo
 					       struct nameidata *new_nd)
 { }
 
+static inline int security_sb_chroot (struct nameidata *nd)
+{
+	return 0;
+}
+
 static inline int security_inode_alloc (struct inode *inode)
 {
 	return 0;
Index: linux-2.6.19.1/security/dummy.c
===================================================================
--- linux-2.6.19.1.orig/security/dummy.c
+++ linux-2.6.19.1/security/dummy.c
@@ -248,6 +248,11 @@ static void dummy_sb_post_pivotroot (str
 	return;
 }
 
+static int dummy_sb_chroot (struct nameidata *nd)
+{
+	return 0;
+}
+
 static int dummy_inode_alloc_security (struct inode *inode)
 {
 	return 0;
@@ -1003,6 +1008,7 @@ void security_fixup_ops (struct security
 	set_to_dummy_if_null(ops, sb_post_addmount);
 	set_to_dummy_if_null(ops, sb_pivotroot);
 	set_to_dummy_if_null(ops, sb_post_pivotroot);
+	set_to_dummy_if_null(ops, sb_chroot);
 	set_to_dummy_if_null(ops, inode_alloc_security);
 	set_to_dummy_if_null(ops, inode_free_security);
 	set_to_dummy_if_null(ops, inode_init_security);
