Index: b/security/apparmor/lsm.c
===================================================================
--- a/security/apparmor/lsm.c
+++ b/security/apparmor/lsm.c
@@ -59,42 +59,6 @@ int apparmor_path_max = 2 * PATH_MAX;
 module_param_named(path_max, apparmor_path_max, int, S_IRUSR);
 MODULE_PARM_DESC(apparmor_path_max, "Maximum pathname length allowed");
 
-#ifndef MODULE
-static int __init aa_getopt_complain(char *str)
-{
-	get_option(&str, &apparmor_complain);
-	return 1;
-}
-__setup("apparmor_complain=", aa_getopt_complain);
-
-static int __init aa_getopt_debug(char *str)
-{
-	get_option(&str, &apparmor_debug);
-	return 1;
-}
-__setup("apparmor_debug=", aa_getopt_debug);
-
-static int __init aa_getopt_audit(char *str)
-{
-	get_option(&str, &apparmor_audit);
-	return 1;
-}
-__setup("apparmor_audit=", aa_getopt_audit);
-
-static int __init aa_getopt_logsyscall(char *str)
-{
-	get_option(&str, &apparmor_logsyscall);
-	return 1;
-}
-__setup("apparmor_logsyscall=", aa_getopt_logsyscall);
-
-static int __init aa_getopt_path_max(char *str)
-{
-	get_option(&str, &apparmor_path_max);
-	return 1;
-}
-__setup("apparmor_path_max=", aa_getopt_path_max);
-#endif
 
 static int aa_reject_syscall(struct task_struct *task, gfp_t flags,
 			     const char *name)
