Index: subdomain/access.sh
===================================================================
--- subdomain.orig/access.sh
+++ subdomain/access.sh
@@ -64,7 +64,7 @@ runchecktest "ACCESS dir rwx (rwx)" pass
 genprofile $dir:$rperm
 runchecktest "ACCESS dir r (r)" pass $dir r
 runchecktest "ACCESS dir rx (r)" pass $dir rx
-runchecktest "ACCESS dir rwx (r)" pass $dir rwx
+runchecktest "ACCESS dir rwx (r)" fail $dir rwx
 
 genprofile $dir:$wxperm
 runchecktest "ACCESS dir x (wx)" pass $dir x
