Index: subdomain/exec_qual.sh
===================================================================
--- subdomain.orig/exec_qual.sh
+++ subdomain/exec_qual.sh
@@ -131,13 +131,13 @@ local_runchecktest "enforce ix case4" pa
 #	  expected behaviour, child should be able to access resource
 
 genprofile $test2:ux $file:$fileperm
-local_runchecktest "enforce ux case1" pass "unconstrained" $test2 $file
+local_runchecktest "enforce ux case1" pass "unconfined" $test2 $file
 
 # case 2: parent profile denies access (should be irrelevant)
 #	  expected behaviour, child should be able to access resource
 
 genprofile $test2:ux
-local_runchecktest "enforce ux case1" pass "unconstrained" $test2 $file
+local_runchecktest "enforce ux case1" pass "unconfined" $test2 $file
 
 # confined parent, exec child with conflicting exec qualifiers
 # that overlap in such away that px is prefered (ix is glob, px is exact
@@ -165,7 +165,7 @@ local_runchecktest "enforce unconfined c
 #	  expected behaviour: child should be able to access resource
 
 removeprofile
-local_runchecktest "enforce unconfined case3" pass "unconstrained" $test2 $file
+local_runchecktest "enforce unconfined case3" pass "unconfined" $test2 $file
 
 # -----------------------------------------------------------------------
 
