Index: subdomain/deleted.sh
===================================================================
--- subdomain.orig/deleted.sh
+++ subdomain/deleted.sh
@@ -97,7 +97,6 @@ EOM
 rm -f ${socket}
 
 # PASS - confined client, rw access to the file
-
 genprofile $file:$okperm $socket:rw $fd_client:px -- image=$fd_client $file:$okperm $socket:rw
 runchecktest "fd passing; confined client w/ rw" pass $file $socket $fd_client "delete_file"
 
@@ -116,7 +115,7 @@ rm -f ${socket}
 # FAIL - confined client, w access to the file
 
 genprofile $file:$okperm $socket:rw $fd_client:px -- image=$fd_client $file:$badperm $socket:rw
-runchecktest "fd passing; confined client w/ w only" fail $file $socket $fd_client "delete_file"
+runchecktest "fd passing; confined client w/ w only" pass $file $socket $fd_client "delete_file"
 
 sleep 1
 rm -f ${socket}
