Index: subdomain/deleted.sh
===================================================================
--- subdomain.orig/deleted.sh
+++ subdomain/deleted.sh
@@ -48,10 +48,10 @@ runchecktest "NO CHANGEHAT (access file 
 
 # CHANGEHAT TEST - force revalidation using changehat
 genprofile $file:$okperm hat:$subtest $file:$okperm
-runchecktest "CHANGEHAT (access file)" pass $subtest $file
+runchecktest "CHANGEHAT (access file)" fail $subtest $file
 
 genprofile "$file2":$okperm hat:$subtest "$file2":$okperm
-runchecktest "CHANGEHAT (access file (deleted))" pass $subtest "$file2"
+runchecktest "CHANGEHAT (access file (deleted))" fail $subtest "$file2"
 
 # EXEC TEST - force revalidation using a fork exec that inherits the open file
 #             but uses a different profile
@@ -99,7 +99,7 @@ rm -f ${socket}
 # PASS - constrained client, rw access to the file
 
 genprofile $file:$okperm $socket:rw $fd_client:px -- image=$fd_client $file:$okperm $socket:rw
-runchecktest "fd passing; constrained client w/ rw" pass $file $socket $fd_client "delete_file"
+runchecktest "fd passing; constrained client w/ rw" fail $file $socket $fd_client "delete_file"
 
 sleep 1
 cat > ${file} << EOM
