Index: subdomain/capabilities.sh
===================================================================
--- subdomain.orig/capabilities.sh
+++ subdomain/capabilities.sh
@@ -74,11 +74,11 @@ syscall_sethostname_args=a.dumb.example.
 syscall_setdomainname_args=dumb.example.com
 syscall_ioperm_args="0 0x3ff"
 syscall_iopl_args=3
-syscall_chroot_args=${tmpdir}
+syscall_chroot_args=${tmpdir}/
 syscall_ptrace_args=sub
 
 # if a testcase requires extra subdomain rules, add them here
-syscall_chroot_extra_entries="/:r ${tmpdir}:r"
+syscall_chroot_extra_entries="/:r ${tmpdir}/:r"
 syscall_ptrace_extra_entries="hat:sub"
 
 testwrapper=changehat_wrapper
Index: subdomain/mult_mount.sh
===================================================================
--- subdomain.orig/mult_mount.sh
+++ subdomain/mult_mount.sh
@@ -48,8 +48,8 @@ file1b=$mp1/file2
 file2a=$mp2/file
 file2b=$mp2/file2
 
-dir1=$mp1/dir
-dir2=$mp2/dir
+dir1=$mp1/dir/
+dir2=$mp2/dir/
 
 mkdirperm=w
 mkdirperm_fail=r
Index: subdomain/readdir.sh
===================================================================
--- subdomain.orig/readdir.sh
+++ subdomain/readdir.sh
@@ -30,12 +30,12 @@ mkdir $dir
 
 # CHDIR TEST
 
-genprofile $dir:$okperm
+genprofile $dir/:$okperm
 
 runchecktest "READDIR" pass $dir
 
 # CHDIR TEST (no perm)
 
-genprofile $dir:$badperm
+genprofile $dir/:$badperm
 
 runchecktest "READDIR (no perm)" fail $dir
Index: subdomain/rename.sh
===================================================================
--- subdomain.orig/rename.sh
+++ subdomain/rename.sh
@@ -24,8 +24,8 @@ bin=$pwd
 
 file1=$tmpdir/file1
 file2=$tmpdir/file2
-dir1=$tmpdir/dir1
-dir2=$tmpdir/dir2
+dir1=$tmpdir/dir1/
+dir2=$tmpdir/dir2/
 
 okfile1perm=rw
 badfile1perm1=r
Index: subdomain/access.sh
===================================================================
--- subdomain.orig/access.sh
+++ subdomain/access.sh
@@ -22,7 +22,7 @@ bin=$pwd
 
 file=$tmpdir/file
 
-dir=$tmpdir/dir
+dir=$tmpdir/dir/
 rperm=r
 rwxperm=rwix
 wxperm=wix
Index: subdomain/mkdir.sh
===================================================================
--- subdomain.orig/mkdir.sh
+++ subdomain/mkdir.sh
@@ -18,7 +18,7 @@ bin=$pwd
 
 . $bin/prologue.inc
 
-dir=$tmpdir/tmpdir
+dir=$tmpdir/tmpdir/
 perms=w
 excess_perms=wl
 badperms=r
Index: subdomain/xattrs.sh
===================================================================
--- subdomain.orig/xattrs.sh
+++ subdomain/xattrs.sh
@@ -40,7 +40,7 @@ bin=$pwd
 
 file=$tmpdir/testfile
 link=$tmpdir/testlink
-dir=$tmpdir/testdir
+dir=$tmpdir/testdir/
 okperm=rw
 badperm=r
 
