Index: subdomain/capabilities.sh
===================================================================
--- subdomain.orig/capabilities.sh
+++ subdomain/capabilities.sh
@@ -74,10 +74,10 @@ syscall_sethostname_args=a.dumb.example.
 syscall_setdomainname_args=dumb.example.com
 syscall_ioperm_args="0 0x3ff"
 syscall_iopl_args=3
-syscall_chroot_args=${tmpdir}
+syscall_chroot_args=${tmpdir}/
 
 # if a testcase requires extra subdomain rules, add them here
-syscall_chroot_extra_entries="/:r ${tmpdir}:r"
+syscall_chroot_extra_entries="/:r ${tmpdir}/:r"
 
 testwrapper=changehat_wrapper
 
Index: subdomain/mult_mount.sh
===================================================================
--- subdomain.orig/mult_mount.sh
+++ subdomain/mult_mount.sh
@@ -48,8 +48,8 @@ file1b=$mp1/file2
 file2a=$mp2/file
 file2b=$mp2/file2
 
-dir1=$mp1/dir
-dir2=$mp2/dir
+dir1=$mp1/dir/
+dir2=$mp2/dir/
 
 mkdirperm=w
 mkdirperm_fail=r
Index: subdomain/readdir.sh
===================================================================
--- subdomain.orig/readdir.sh
+++ subdomain/readdir.sh
@@ -30,12 +30,12 @@ mkdir $dir
 
 # CHDIR TEST
 
-genprofile $dir:$okperm
+genprofile $dir/:$okperm
 
 runchecktest "READDIR" pass $dir
 
 # CHDIR TEST (no perm)
 
-genprofile $dir:$badperm
+genprofile $dir/:$badperm
 
 runchecktest "READDIR (no perm)" fail $dir
Index: subdomain/rename.sh
===================================================================
--- subdomain.orig/rename.sh
+++ subdomain/rename.sh
@@ -24,8 +24,8 @@ bin=$pwd
 
 file1=$tmpdir/file1
 file2=$tmpdir/file2
-dir1=$tmpdir/dir1
-dir2=$tmpdir/dir2
+dir1=$tmpdir/dir1/
+dir2=$tmpdir/dir2/
 
 okfile1perm=rw
 badfile1perm1=r
Index: subdomain/access.sh
===================================================================
--- subdomain.orig/access.sh
+++ subdomain/access.sh
@@ -22,7 +22,7 @@ bin=$pwd
 
 file=$tmpdir/file
 
-dir=$tmpdir/dir
+dir=$tmpdir/dir/
 rperm=r
 rwxperm=rwix
 wxperm=wix
Index: subdomain/mkdir.sh
===================================================================
--- subdomain.orig/mkdir.sh
+++ subdomain/mkdir.sh
@@ -18,7 +18,7 @@ bin=$pwd
 
 . $bin/prologue.inc
 
-dir=$tmpdir/tmpdir
+dir=$tmpdir/tmpdir/
 perms=w
 excess_perms=wl
 badperms=r
Index: subdomain/xattrs.sh
===================================================================
--- subdomain.orig/xattrs.sh
+++ subdomain/xattrs.sh
@@ -40,7 +40,7 @@ bin=$pwd
 
 file=$tmpdir/testfile
 link=$tmpdir/testlink
-dir=$tmpdir/testdir
+dir=$tmpdir/testdir/
 okperm=rw
 badperm=r
 
Index: subdomain/openat.sh
===================================================================
--- subdomain.orig/openat.sh
+++ subdomain/openat.sh
@@ -22,43 +22,44 @@ bin=$pwd
 
 . $bin/prologue.inc
 
+dir=${tmpdir}/
 subdir=deleteme
 file=${subdir}/file
-filepath=${tmpdir}/${file}
+filepath=${dir}${file}
 okperm=rw
 badperm1=r
 badperm2=w
 
-mkdir ${tmpdir}/${subdir}
+mkdir ${dir}/${subdir}
 
 # PASS UNCONFINED
-runchecktest "OPENAT unconfined RW (create) " pass $tmpdir $file
+runchecktest "OPENAT unconfined RW (create) " pass $dir $file
 
 # PASS TEST (the file shouldn't exist, so open should create it
 rm -f ${filepath}
-genprofile ${tmpdir}:r ${filepath}:$okperm
-runchecktest "OPENAT RW (create) " pass $tmpdir $file
+genprofile ${dir}:r ${filepath}:$okperm
+runchecktest "OPENAT RW (create) " pass $dir $file
 
 # PASS TEST
-genprofile ${tmpdir}:r ${filepath}:$okperm
-runchecktest "OPENAT RW" pass $tmpdir $file
+genprofile ${dir}:r ${filepath}:$okperm
+runchecktest "OPENAT RW" pass $dir $file
 
 # FAILURE TEST (1)
-genprofile ${tmpdir}:r ${filepath}:$badperm1
-runchecktest "OPENAT R" fail $tmpdir $file
+genprofile ${dir}:r ${filepath}:$badperm1
+runchecktest "OPENAT R" fail $dir $file
 
 # FAILURE TEST (2)
-genprofile ${tmpdir}:r ${filepath}:$badperm2
-runchecktest "OPENAT W" fail $tmpdir $file
+genprofile ${dir}:r ${filepath}:$badperm2
+runchecktest "OPENAT W" fail $dir $file
 
 # FAILURE TEST (3)
-genprofile ${tmpdir}:r ${filepath}:$badperm1 cap:dac_override
-runchecktest "OPENAT R+dac_override" fail $tmpdir $file
+genprofile ${dir}:r ${filepath}:$badperm1 cap:dac_override
+runchecktest "OPENAT R+dac_override" fail $dir $file
 
 # FAILURE TEST (4)
 # This is testing for bug: https://bugs.wirex.com/show_bug.cgi?id=2885
 # When we open O_CREAT|O_RDWR, we are (were?) allowing only write access
 # to be required.
 rm -f ${filepath}
-genprofile ${tmpdir}:r ${filepath}:$badperm2
-runchecktest "OPENAT W (create)" fail $tmpdir $file
+genprofile ${dir}:r ${filepath}:$badperm2
+runchecktest "OPENAT W (create)" fail $dir $file
