Index: subdomain/mult_mount.sh
===================================================================
--- subdomain.orig/mult_mount.sh
+++ subdomain/mult_mount.sh
@@ -79,17 +79,17 @@ genprofile $dir2:$mkdirperm
 runchecktest "OVERLAP MKDIR PASS2" pass mkdir $dir2
 cleandir $dir2
 
+# MKDIR FAIL TESTS
 genprofile $dir1:$mkdirperm
 
-runchecktest "OVERLAP MKDIR PASS3" pass mkdir $dir2
+runchecktest "OVERLAP MKDIR PASS3" fail mkdir $dir2
 cleandir $dir2
 
 genprofile $dir2:$mkdirperm
 
-runchecktest "OVERLAP MKDIR PASS4" pass mkdir $dir1
+runchecktest "OVERLAP MKDIR PASS4" fail mkdir $dir1
 cleandir $dir1
 
-# MKDIR FAIL TESTS
 get_current_logposition
 genprofile $dir1:$mkdirperm_fail
 runchecktest "OVERLAP MKDIR FAIL1" fail mkdir $dir1
@@ -114,11 +114,11 @@ runchecktest "OVERLAP LINK PASS1a" pass 
 rm -f $file1b
 
 genprofile $file1a:$linkperm $file2b:$linkperm
-runchecktest "OVERLAP LINK PASS2a" pass $file1a $file1b
+runchecktest "OVERLAP LINK PASS2a" fail $file1a $file1b
 rm -f $file1b
 
 genprofile $file2a:$linkperm $file1b:$linkperm
-runchecktest "OVERLAP LINK PASS3a" pass $file1a $file1b
+runchecktest "OVERLAP LINK PASS3a" fail $file1a $file1b
 rm -f $file1b
 
 # src lacks link perm, ok as long as dest has linkperm
@@ -127,11 +127,11 @@ runchecktest "OVERLAP LINK PASS1b" pass 
 rm -f $file1b
 
 genprofile $file1a:$readperm $file2b:$linkperm
-runchecktest "OVERLAP LINK PASS2b" pass $file1a $file1b
+runchecktest "OVERLAP LINK PASS2b" fail $file1a $file1b
 rm -f $file1b
 
 genprofile $file2a:$readperm $file1b:$linkperm
-runchecktest "OVERLAP LINK PASS3b" pass $file1a $file1b
+runchecktest "OVERLAP LINK PASS3b" fail $file1a $file1b
 rm -f $file1b
 
 
