Index: linux-2.6.18.6/security/apparmor/apparmor.h
===================================================================
--- linux-2.6.18.6.orig/security/apparmor/apparmor.h
+++ linux-2.6.18.6/security/apparmor/apparmor.h
@@ -210,7 +210,7 @@ typedef int (*aa_iter) (struct subdomain
  */
 struct aa_path_data {
 	struct dentry *root, *dentry;
-	struct mnt_namespace *mnt_namespace;
+	struct namespace *namespace;
 	struct list_head *head, *pos;
 	int errno;
 };
Index: linux-2.6.18.6/security/apparmor/inline.h
===================================================================
--- linux-2.6.18.6.orig/security/apparmor/inline.h
+++ linux-2.6.18.6/security/apparmor/inline.h
@@ -10,7 +10,7 @@
 #ifndef __INLINE_H
 #define __INLINE_H
 
-#include <linux/mnt_namespace.h>
+#include <linux/namespace.h>
 
 static inline int __aa_is_confined(struct subdomain *sd)
 {
@@ -323,8 +323,8 @@ static inline void __aa_path_begin(struc
 {
 	data->dentry = dentry;
 	data->root = dget(rdentry->d_sb->s_root);
-	data->mnt_namespace = current->nsproxy->mnt_ns;
-	data->head = &data->mnt_namespace->list;
+	data->namespace = current->namespace;
+	data->head = &data->namespace->list;
 	data->pos = data->head->next;
 	prefetch(data->pos->next);
 	data->errno = 0;
