Index: linux-2.6.16.29/security/apparmor/lsm.c
===================================================================
--- linux-2.6.16.29.orig/security/apparmor/lsm.c
+++ linux-2.6.16.29/security/apparmor/lsm.c
@@ -176,9 +176,9 @@ static int apparmor_netlink_send(struct 
 	return cap_netlink_send(sk, skb);
 }
 
-static int apparmor_netlink_recv(struct sk_buff *skb, int cap)
+static int apparmor_netlink_recv(struct sk_buff *skb)
 {
-	return cap_netlink_recv(skb, cap);
+	return cap_netlink_recv(skb);
 }
 
 static void apparmor_bprm_apply_creds(struct linux_binprm *bprm, int unsafe)
