Index: linux-2.6.19/fs/namei.c
===================================================================
--- linux-2.6.19.orig/fs/namei.c
+++ linux-2.6.19/fs/namei.c
@@ -1931,7 +1931,7 @@ int vfs_mkdir(struct inode *dir, struct 
 		return -EPERM;
 
 	mode &= (S_IRWXUGO|S_ISVTX);
-	error = security_inode_mkdir(dir, dentry, mode);
+	error = security_inode_mkdir(dir, mnt, dentry, mode);
 	if (error)
 		return error;
 
Index: linux-2.6.19/include/linux/security.h
===================================================================
--- linux-2.6.19.orig/include/linux/security.h
+++ linux-2.6.19/include/linux/security.h
@@ -307,6 +307,7 @@ struct request_sock;
  *	Check permissions to create a new directory in the existing directory
  *	associated with inode strcture @dir. 
  *	@dir containst the inode structure of parent of the directory to be created.
+ *	@mnt is the vfsmount of the new dentry (may be NULL)
  *	@dentry contains the dentry structure of new directory.
  *	@mode contains the mode of new directory.
  *	Return 0 if permission is granted.
@@ -1221,7 +1222,8 @@ struct security_operations {
 	int (*inode_unlink) (struct inode *dir, struct dentry *dentry);
 	int (*inode_symlink) (struct inode *dir,
 	                      struct dentry *dentry, const char *old_name);
-	int (*inode_mkdir) (struct inode *dir, struct dentry *dentry, int mode);
+	int (*inode_mkdir) (struct inode *dir, struct vfsmount *mnt,
+			    struct dentry *dentry, int mode);
 	int (*inode_rmdir) (struct inode *dir, struct dentry *dentry);
 	int (*inode_mknod) (struct inode *dir, struct dentry *dentry,
 	                    int mode, dev_t dev);
@@ -1658,12 +1660,13 @@ static inline int security_inode_symlink
 }
 
 static inline int security_inode_mkdir (struct inode *dir,
+					struct vfsmount *mnt,
 					struct dentry *dentry,
 					int mode)
 {
 	if (unlikely (IS_PRIVATE (dir)))
 		return 0;
-	return security_ops->inode_mkdir (dir, dentry, mode);
+	return security_ops->inode_mkdir (dir, mnt, dentry, mode);
 }
 
 static inline int security_inode_rmdir (struct inode *dir,
@@ -2379,6 +2382,7 @@ static inline int security_inode_symlink
 }
 
 static inline int security_inode_mkdir (struct inode *dir,
+					struct vfsmount *mnt,
 					struct dentry *dentry,
 					int mode)
 {
Index: linux-2.6.19/security/dummy.c
===================================================================
--- linux-2.6.19.orig/security/dummy.c
+++ linux-2.6.19/security/dummy.c
@@ -287,8 +287,8 @@ static int dummy_inode_symlink (struct i
 	return 0;
 }
 
-static int dummy_inode_mkdir (struct inode *inode, struct dentry *dentry,
-			      int mask)
+static int dummy_inode_mkdir (struct inode *inode, struct vfsmount *mnt,
+			      struct dentry *dentry, int mask)
 {
 	return 0;
 }
Index: linux-2.6.19/security/selinux/hooks.c
===================================================================
--- linux-2.6.19.orig/security/selinux/hooks.c
+++ linux-2.6.19/security/selinux/hooks.c
@@ -2160,7 +2160,8 @@ static int selinux_inode_symlink(struct 
 	return may_create(dir, dentry, SECCLASS_LNK_FILE);
 }
 
-static int selinux_inode_mkdir(struct inode *dir, struct dentry *dentry, int mask)
+static int selinux_inode_mkdir(struct inode *dir, struct vfsmount *mnt,
+			       struct dentry *dentry, int mask)
 {
 	return may_create(dir, dentry, SECCLASS_DIR);
 }
