---
 security/apparmor/lsm.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/security/apparmor/lsm.c
+++ b/security/apparmor/lsm.c
@@ -205,7 +205,7 @@ static int apparmor_ptrace(struct task_s
 
 			error = aa_may_ptrace(cxt, child_cxt ?
 						   child_cxt->profile : NULL);
-			if (PROFILE_COMPLAIN(cxt->profile)) {
+			if (error && PROFILE_COMPLAIN(cxt->profile)) {
 				struct aa_audit sa;
 				memset(&sa, 0, sizeof(sa));
 				sa.operation = "ptrace";
