---
 security/apparmor/lsm.c |   28 ----------------------------
 1 file changed, 28 deletions(-)

--- a/security/apparmor/lsm.c
+++ b/security/apparmor/lsm.c
@@ -157,20 +157,6 @@ static int param_set_aa_enabled(const ch
 	return 0;
 }
 
-static int aa_reject_syscall(struct task_struct *task, gfp_t flags,
-			     const char *name)
-{
-	struct aa_profile *profile = aa_get_profile(task);
-	int error = 0;
-
-	if (profile) {
-		error = aa_audit_syscallreject(profile, flags, name);
-		aa_put_profile(profile);
-	}
-
-	return error;
-}
-
 static int apparmor_ptrace(struct task_struct *parent,
 			   struct task_struct *child)
 {
@@ -294,17 +280,6 @@ static int apparmor_bprm_secureexec(stru
 	return ret;
 }
 
-static int apparmor_sb_mount(char *dev_name, struct path *path, char *type,
-			      unsigned long flags, void *data)
-{
-	return aa_reject_syscall(current, GFP_KERNEL, "mount");
-}
-
-static int apparmor_umount(struct vfsmount *mnt, int flags)
-{
-	return aa_reject_syscall(current, GFP_KERNEL, "umount");
-}
-
 static int apparmor_inode_mkdir(struct inode *dir, struct dentry *dentry,
 				struct vfsmount *mnt, int mask)
 {
@@ -977,9 +952,6 @@ struct security_operations apparmor_ops
 	.bprm_set_security =		apparmor_bprm_set_security,
 	.bprm_secureexec =		apparmor_bprm_secureexec,
 
-	.sb_mount =			apparmor_sb_mount,
-	.sb_umount =			apparmor_umount,
-
 	.inode_mkdir =			apparmor_inode_mkdir,
 	.inode_rmdir =			apparmor_inode_rmdir,
 	.inode_create =			apparmor_inode_create,
